src/v3/core/address.test.js

```diff
@@ -58,5 +58,142 @@ describe("core/address", () => {
+        expect(() => {
+          FooAddress.assertValid("F/things\0", "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+      it("rejects a string that does not end with a separator", () => {
+        const address = FooAddress.fromParts(["hello"]);
+        const truncated = address.substring(0, address.length - 2);
+        expect(() => {
+          FooAddress.assertValid(truncated, "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+      it("rejects junk", () => {
+        expect(() => {
+          FooAddress.assertValid("junk", "widget");
+        }).toThrow("widget: expected FooAddress, got:");
+      });
+    });
+
+    describe("assertValidParts", () => {
+      const {FooAddress} = makeModules();
+      it("rejects `undefined`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(undefined, "widget");
+        }).toThrow("widget: expected array of parts, got: undefined");
+      });
+      it("rejects `null`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(null, "widget");
+        }).toThrow("widget: expected array of parts, got: null");
+      });
+      it("rejects an array containing `undefined`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(["hello", undefined, "world"], "widget");
+        }).toThrow(
+          "widget: expected array of parts, got undefined in: " +
+            '["hello",undefined,"world"]'
+        );
+      });
+      it("rejects an array containing `null`", () => {
+        expect(() => {
+          // $ExpectFlowError
+          FooAddress.assertValidParts(["hello", null, "world"], "widget");
+        }).toThrow(
+          "widget: expected array of parts, got null in: " +
+            '["hello",null,"world"]'
+        );
+      });
+      it("rejects an array with a string containing a NUL character", () => {
+        expect(() => {
+          FooAddress.assertValidParts(["hello", "n\0o", "world"], "widget");
+        }).toThrow(
+          "widget: part contains NUL character: " +
+            '"n\\u0000o" in ["hello","n\\u0000o","world"]'
+        );
+      });
+    });
+
+    describe("fromParts", () => {
+      const {FooAddress, BarAddress} = makeModules();
+      it("encodes the address kind for the empty address", () => {
+        const foo = FooAddress.fromParts([]);
+        const bar = BarAddress.fromParts([]);
+        expect(foo).not.toEqual(bar);
+      });
+      it("encodes the address kind for a normal address", () => {
+        const foo = FooAddress.fromParts(["hello", "world"]);
+        const bar = BarAddress.fromParts(["hello", "world"]);
+        expect(foo).not.toEqual(bar);
+      });
+    });
+
+    describe("toParts", () => {
+      const {FooAddress, BarAddress} = makeModules();
+
+      // We use this next test as a proxy for fully correct validation,
```

Comment:
Is explicit mocking feasible?

Reply:
Did you want to mock `assertValid` and `assertValidParts` to check that
they’re called? That’s reasonable, but definitely increases the
fragility of the tests. For instance, suppose that the implementation
were changed to

```js
function _assertValid(…) { … }
function assertValid(...args) {
  if (NODE_ENV !== "production") {
    return assertValid(...args);
  }
  // or, do some other preprocessing, like pre-compute a default value
  // of `what`, or catch and rethrow errors adding some information to
  // the message, or…
}
function toParts(address) {
  _assertValid(address);
  // …
}
```

Then our mock that `assertValid` is called would fail, because only the
internal function `_assertValid` is called.

I definitely prefer not to mock when we don’t have to, as it increases
complexity and means that the code that actually runs in production is
not the code that you’re testing. Sometimes it’s obviously the right
thing to do, but I don’t think that that’s the case here.

I’ll merge as-is; if I misunderstood or you think that mocking here is a
really great idea, then I’m happy to change this.